### PR TITLE
Closes #2142 Supersede `get_summary_records()` and enhance `derive_summary_records()`

### LINKED ISSUE #2142: Feature Request:  Enhance `derive_summary_records()` and supersede `get_summary_records()`
### Feature Idea

`derive_summary_records()` should be updated such that it can work with `dataset_add`, `dataset_ref`, `missing values` while deprecating the `analysis_var` and `summary_fun` arguments.

`get_summary_records()` can be superseded with the addition of `dataset_add` per conversation [here](https://github.com/pharmaverse/admiral/pull/2124#discussion_r1336773356)

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by zdz2101**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -38,6 +38,7 @@ LazyData: true
 Roxygen: list(markdown = TRUE)
 RoxygenNote: 7.2.3
 Depends: R (>= 4.0)
+Remotes: pharmaverse/admiraldev
```
Comment (zdz2101) [inline DESCRIPTION:41]: @ddsjoberg is this all I would have to do? or does I need to remove the imports line too?

Comment (ddsjoberg) [reply DESCRIPTION:41]: You'll want to make one small change to include the min version number required.

```
Remotes: pharmaverse/admiraldev (>= 0.5.0.9000)
```

Then if anyone does any dev work, like redocumenting, and they don't have this version installed, they'll get a note that they need to install the updated version.

File: R/get_summary_records.R [hunk 2]
```diff
@@ -64,9 +71,9 @@
 #'
 #' @return A data frame of derived records.
 #'
-#' @family der_gen
+#' @family superseded
 #'
-#' @keywords der_gen
+#' @keywords internal
```
Comment (bms63) [inline R/get_summary_records.R:76]: I made a superseded section on the reference page.

```suggestion
#' @keywords superseded
```

**Review by bms63 — COMMENTED** [review]
Do we not use this is any vignettes?

File: NEWS.md [hunk 3]
```diff
@@ -12,6 +12,8 @@
 were enhanced such that more than one summary variable can be derived, e.g.,
 `AVAL` as the sum and `ADT` as the maximum of the contributing records. (#1792)
 
+- `derive_summary_records()` was enhanced with the following optional arguments: `dataset_add`, `dataset_ref`, `missing_values`. These arguments respectively, generate summary variables from additional datasets, retain/add specific records that and impute user-defined missing values. (#2142)
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply NEWS.md]: fixed

File: R/get_summary_records.R [hunk 4]
```diff
@@ -1,6 +1,13 @@
 #' Create Summary Records
 #'
 #' @description
+#'
+#' `r lifecycle::badge("superseded")`
+#'
+#' Development on `get_summary_records()` is complete, and for new code we recommend
+#' switching to using the `dataset_add` argument in `derive_summary_records()`,
+#' which is easier to use, has more features, and still under active development.
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/get_summary_records.R]: fixed

File: R/get_summary_records.R (hunk 2 above)
Comment (zdz2101) [reply R/get_summary_records.R]: fixed

**Comment by zdz2101** [comment]
> Do we not use this is any vignettes?

We do, but @bundfussr already made the corresponding updates in https://github.com/pharmaverse/admiral/pull/2124

File: R/derive_summary_records.R [hunk 5]
```diff
@@ -125,18 +165,33 @@
 #' ) %>%
 #'   arrange(USUBJID, AVISIT)
 derive_summary_records <- function(dataset,
```
Comment (bundfussr) [inline R/derive_summary_records.R:167]: Default should be `NULL`.

File: R/derive_summary_records.R [hunk 6]
```diff
@@ -125,18 +165,33 @@
...
+                                   dataset_add = NULL,
```
Comment (bundfussr) [inline R/derive_summary_records.R:168]: `dataset_add` should be mandatory to be consistent with the other functions.

File: R/derive_summary_records.R [hunk 7]
```diff
@@ -125,18 +165,33 @@
...
+                                   dataset_ref = NULL,
                                    by_vars,
                                    filter = NULL,
```
Comment (bundfussr) [inline R/derive_summary_records.R:171]: For consistency we should rename `filter` to `filter_add`.

File: R/derive_summary_records.R [hunk 8]
```diff
@@ -125,18 +165,33 @@
...
                                    analysis_var,
                                    summary_fun,
-                                   set_values_to) {
+                                   set_values_to = NULL,
```
Comment (bundfussr) [inline R/derive_summary_records.R:174]: Should be mandatory, i.e., no default value.

File: R/derive_summary_records.R [hunk 9]
```diff
@@ -125,18 +165,33 @@
...
+                                   missing_values = NULL) {
   assert_vars(by_vars)
   filter <- assert_filter_cond(enexpr(filter), optional = TRUE)
   assert_data_frame(
     dataset,
-    required_vars = by_vars
+    required_vars = expr_c(by_vars)
```
Comment (bundfussr) [inline R/derive_summary_records.R:180]: Why do we need `expr_c()`?

File: R/derive_summary_records.R [hunk 10]
```diff
@@ -149,14 +204,47 @@ derive_summary_records <- function(dataset,
...
+  if (!is.null(missing_values)) {
+    update_missings <- map2(
+      syms(names(missing_values)),
+      missing_values,
+      ~ expr(if_else(is.na(!!.x), !!.y, !!.x))
     )
+    names(update_missings) <- names(missing_values)
+    summary_records <- summary_records %>%
+      mutate(!!!update_missings)
+  }
```
Comment (bundfussr) [inline R/derive_summary_records.R:226]: I think `missing_values` should be applied to `new_ref_obs` below.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_summary_records.R [hunk 11]
```diff
@@ -29,6 +41,34 @@
...
+#'     set_values_to = exprs(
+#'       AVAL = sum(AVAL),
+#'       DTYPE = "AVERAGE",
+#'     )
+#'   ```
+#'
+#' @param missing_values Values for missing summary values
+#'
+#'   For observations of the input dataset (`dataset`) or (`dataset_add`) which do not have an
+#'   complete mapping defined by the summarization defined in `set_values_to`.  Only variables
```
Comment (bms63) [inline R/derive_summary_records.R:66]:
```suggestion
#'   For observations of the input dataset (`dataset`) or (`dataset_add`) which do not have a
#'   complete mapping defined by the summarization defined in `set_values_to`.  Only variables
```

File: R/derive_summary_records.R (hunk 10 above)
Comment (bundfussr) [reply R/derive_summary_records.R:244]: The `missing_values` argument is intended to define the values for by groups which are not in `dataset_add`. For example, if you derive the total dose, you could use `missing_values = exprs(AVAL = 0)` to set the total dose to zero for patients without exposure records.

Comment (zdz2101) [reply R/derive_summary_records.R:244]: Got it, that makes much more sense

File: R/derive_summary_records.R (hunk 6 above)
Comment (bundfussr) [reply R/derive_summary_records.R]: Yes, there are some implications but if implement `dataset_add` as optional, we should do it in all other functions as well. And this could have much more implications.
It's a pity that we do not have more time.

File: tests/testthat/test-derive_summary_records.R [hunk 12]
```diff
@@ -143,3 +146,91 @@ test_that("derive_summary_records Test 4: deprecation warning for analysis_var a
...
+    "2", 2,
+  )
+
+  expected_output <- bind_rows(
+    input,
+    tibble::tribble(
+      ~subj, ~visit,       ~aval,         ~type,
+      "1",        1,          11,     "AVERAGE",
+      "1",        2,          11,     "AVERAGE",
+      "2",        1,      999999,     "MISSING",
```
Comment (zdz2101) [inline tests/testthat/test-derive_summary_records.R:214]: missing_values test

File: NEWS.md [hunk 13]
```diff
@@ -17,6 +17,8 @@ character vector (`'--DTC'`), was imputed. (#2146)
 were enhanced such that more than one summary variable can be derived, e.g.,
 `AVAL` as the sum and `ADT` as the maximum of the contributing records. (#1792)
 
+- `derive_summary_records()` was enhanced with the following optional arguments: `dataset_add`, `dataset_ref`, `missing_values`. These arguments respectively, generate summary variables from additional datasets, retain/add specific records from a reference dataset, and impute user-defined missing values. (#2142)
+
```
Comment (bundfussr) [inline NEWS.md:21]: We should add that
- `dataset` is optional now (in `derive_summary_records()` and `derive_param_exposure()`)
- `dataset_add` was added to `derive_param_exposure()`

File: R/derive_summary_records.R [hunk 14]
```diff
@@ -116,6 +174,7 @@
 #' # by group
 #' derive_summary_records(
 #'   adeg,
+#'   dataset_add = adeg,
 #'   by_vars = exprs(USUBJID, PARAM, AVISIT),
 #'   filter = n() > 2,
```
Comment (bundfussr) [inline R/derive_summary_records.R:179]: Should be `filter_add`.

File: R/derive_param_exposure.R [hunk 15]
```diff
@@ -8,13 +8,40 @@
 #'   `PARAMCD` is expected as well,
 #'   + Either `ASTDTM` and `AENDTM` or `ASTDT` and `AENDT` are also expected.
```
Comment (bundfussr) [inline R/derive_param_exposure.R:9]: This and `analysis_var` should be moved to `dataset_add`.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_param_exposure.R [hunk 16]
```diff
@@ -6,15 +6,42 @@
 #' @param dataset
 #'   `r roxygen_param_dataset(expected_vars = c("by_vars", "analysis_var"))`
 #'   `PARAMCD` is expected as well,
```
Comment (bundfussr) [inline R/derive_param_exposure.R:8]: `analysis_var` and `PARAMCD` should be removed.

File: R/derive_param_exposure.R [hunk 17]
```diff
@@ -6,15 +6,42 @@
...
-#'   + Either `ASTDTM` and `AENDTM` or `ASTDT` and `AENDT` are also expected.
 #'
-#' @param filter Filter condition
+#' @param dataset_add Additional dataset
 #'
-#'   The specified condition is applied to the input dataset before deriving the
-#'   new parameter, i.e., only observations fulfilling the condition are taken
-#'   into account.
+#'   The variables specified for `by_vars`, `analysis_var`,
+#'    + alongside either `ASTDTM` and `AENDTM` or `ASTDT` and `AENDT` are also expected.
```
Comment (bundfussr) [inline R/derive_param_exposure.R:13]: I think the "+" should be removed (the rendered versions looks strange).

And `PARAMCD` should be added.

**Review by bundfussr — CHANGES_REQUESTED** [review]

**Review by bundfussr — APPROVED** [review]
